src/subsystem.lisp

```diff
@@ -0,0 +1,119 @@
+  "How many qubits does the subsystem SS represent?"
+  (logcount (subsystem-qubits ss)))
+
+(defmethod print-object ((ss subsystem) stream)
+  (print-unreadable-object (ss stream :type t :identity nil)
+    (let ((qubits (subsystem-qubits ss)))
+      (format stream "~Dq" (subsystem-num-qubits ss))
+      (when (plusp qubits)
+        (format stream ":")
+        (loop :for i :below (integer-length (subsystem-qubits ss))
+              :when (logbitp i qubits)
+                :do (format stream " ~D" i))))))
+
+(defun subsystem-contains-qubit-p (ss qubit)
+  "Does the subsystem SS contain the qubits QUBIT?"
+  (logbitp qubit (subsystem-qubits ss)))
+
+(defun subsystem-contains-qubits-p (ss qubit-set)
+  "Does the subsystem SS contain the qubits designated by the bit-set QUBIT-SET?"
+  (zerop (logandc1 (subsystem-qubits ss) qubit-set)))
+
+(defun subsystem-physical-to-logical-qubit (ss phys-qubit)
+  "What logical qubit numbner does the physical qubit PHYS-QUBIT have in the subsystem SS?
+
+(The  \"logical qubit\" is the relative position of the qubit in the system.)"
+  (and (subsystem-contains-qubit-p ss phys-qubit)
+       (logcount (ldb (byte phys-qubit 0) (subsystem-qubits ss)))))
+
+(defun %collapse-integer-by-mask (integer mask)
+  "Let INTEGER and MASK be integers of the same bit-length. Remove the bits of INTEGER that correspond to the zeros of MASK."
+  (loop :with collapsed := 0
+        :with j := 0
+        :for i :below (integer-length mask)
+        :when (logbitp i mask)
+          :do (setf collapsed (dpb (ldb (byte 1 i) integer)
+                                   (byte 1 j) collapsed))
+              (incf j)
+        :finally (return collapsed)))
+
+;;; Maybe a future TODO:
+;;;
+;;;    - Allow JOIN-SUBSYSTEMS to take any number of args.
+;;;
+;;;    - Allow EJECT-QUBIT-FROM-SUBSYSTEM to take many qubits.
+
+(defun join-subsystems (ss1 ss2)
+  "Join two (disjoint) qubit subsystems SS1 and SS2 into a larger qubit subsystem."
+  (let* ((q1 (subsystem-qubits ss1))
+         (q2 (subsystem-qubits ss2))
+         (new-qubits (logior q1 q2))
+         (collapsed-q1 (%collapse-integer-by-mask q1 new-qubits))
+         (collapsed-q2 (%collapse-integer-by-mask q2 new-qubits)))
+    (unless (zerop (logand q1 q2))
+      (error "The subsystems can't be joined because they share qubits."))
+    (let* ((new-state-size (expt 2 (logcount new-qubits)))
+           (new-state (make-lisp-cflonum-vector new-state-size)))
+      (loop :with ss1-state := (subsystem-state ss1)
+            :with ss2-state := (subsystem-state ss2)
+            :for addr :below new-state-size
+            :do (setf (aref new-state addr)
+                      (* (aref ss1-state (%collapse-integer-by-mask addr collapsed-q1))
+                         (aref ss2-state (%collapse-integer-by-mask addr collapsed-q2)))))
+      (make-subsystem :qubits new-qubits
+                      :state new-state))))
+
+(defun eject-qubit-from-subsystem (ss qubit)
```

Comment:
I don't think this is right.

Reply:
I suppose it depends on what the intended semantics are. My naive interpretation of all of this is that "ejecting" a qubit should generically result in a mixed state, unless the original state factors as a tensor project of the qubit-to-be-ejected and the rest of the subsystem.

Reply:
The intent is for when the qubit is factored out, usually when an ancilla is temporarily used in a compute-extract-uncompute operation. But I didn't know off the top of my head an easy way to verify this.

Reply:
Here's a version that I think is morally correct (but I didn't try to debug mine—I haven't even run it!):
```
(defun eject-qubit-from-subsystem (ss qubit)
  "Eject the qubit QUBIT from the subsystem SS."
  (unless (subsystem-contains-qubit-p ss qubit)
    (return-from eject-qubit-from-subsystem ss))
  (let ((old-state (subsystem-state ss))
        (new-state (make-lisp-cflonum-vector (/ (length old-state) 2)))
        (small-state (make-lisp-cflonum-vector 2))
        index-shift rescale)
    ;; calculate the ancilla state. for old-state assumed to look like
    ;;    a0 b0 |00> + a0 b1 |01> + ... + a0 bn |0n> + a1 b0 |10> + ...
    ;; this loop seeks the largest ai bj, hence largest ai and bj individually,
    ;; which tamps down on numerical error incurred from dividing by near-zero.
    (let ((abs 0d0))
      (dotimes (j (length old-state))
        (let ((small-index (ldb (byte qubit 1) j)))
          (when (> (abs (aref old-state j)) abs)
            (setf abs (abs (aref old-state j))
                  (aref small-state small-index) (aref old-state j)
                  (aref small-state (- 1 small-index)) (aref old-state (xor j (ash 1 qubit))))))))
    ;; small-state looks like a0 bj |0> + a1 bj |1> for some j. normalize out bj
    (setf small-state (normalize-wavefunction small-state))
    ;; calculate the factorized state. still assuming old-state as above,
    ;; we pick the larger al of a0, a1 ...
    (let* ((index-shift (if (> (abs (aref small-state 0) (aref small-state 1)))
                            0 1))
           (rescale (aref small-state index-shift))
           (ratio (aref small-state (- 1 index-shift)) (/ (aref small-state index-shift)))))
    (dotimes (j (length old-state))
      (cond
       ;; ... then examine al b0 |l0> + ... al bn |ln>, and scale out al.
       ((= index-shift (ldb (byte 1 qubit) j))
        (setf (aref new-state (eject-bit qubit j)) (/ (aref old-state j) rescale)))
       ((/= index-shift (ldb (byte 1 qubit) j))
        ;; also, we should be able to infer the value of a(1-l) bj |(1-j)j> in old-state
        ;; from a(1-l) / al * al bj. if this ever fails, the ancilla is actually entangled.
        (assert
         (double= (aref old-state j)
                  (* ratio (aref old-state (xor j (ash 1 qubit)))))))))
    (make-subsystem :qubits (eject-bit (subsystem-qubits ss) qubit)
                    :state new-state)))
```
This version just assumes that the ancilla is unentangled from the bulk. If the ancilla is in a specific state—like, say, the ground state—then you can simplify this much further:
```
(defun eject-qubit-from-subsystem (ss qubit)
  "Eject the qubit QUBIT from the subsystem SS."
  (unless (subsystem-contains-qubit-p ss qubit)
    (return-from eject-qubit-from-subsystem ss))
  (let ((old-state (subsystem-state ss))
        (new-state (make-lisp-cflonum-vector (/ (length old-state) 2))))
    (dotimes (j (length old-state))
      (cond
       ((zerop (ldb (byte 1 qubit) j))
        (setf (aref new-state (eject-bit qubit j))
              (aref old-state j)))
       (t
        (assert (double= 0d0 (aref old-state j))))))
    (make-subsystem :qubits (eject-bit (subsystem-qubits ss) qubit)
                    :state new-state)))
```